src/pq/connection.cr

```diff
@@ -270,6 +271,69 @@ module PQ
       end
     end
 
+    struct SamlContext
+      property client_first_msg : String
+      property client_first_msg_size : Int32
+
+      def initialize(@password : String)
+        @client_nonce = Random::Secure.urlsafe_base64(18)
+        @client_first_msg = "n,,n=,r=#{@client_nonce}"
+        @client_first_msg_size = @client_first_msg.bytesize
+      end
+
+      def generate_client_final_message(body)
+        server_first_msg = String.new(body)
+        params = server_first_msg.split(',')
+        r = params.find { |p| p[0] == 'r' }.not_nil![2..-1]
+        s = params.find { |p| p[0] == 's' }.not_nil![2..-1]
+        i = params.find { |p| p[0] == 'i' }.not_nil![2..-1].to_i
+        raise ConnectionError.new("SASL: scram server nonce does not start with client nonce") unless r.starts_with?(@client_nonce)
+
+        client_final_msg_without_proof = "c=biws,r=#{r}"
+        salted_pass = OpenSSL::PKCS5.pbkdf2_hmac(@password, Base64.decode(s), i, algorithm: OpenSSL::Algorithm::SHA256, key_size: 32)
```

Comment:
@will The password needs to be normalized with SASLPrep (well, PostgreSQL flavored SASLPrep), otherwise some valid passwords will fail. Here is an example of how to do it:

https://github.com/MagicStack/asyncpg/blob/master/asyncpg/protocol/scram.pyx#L263

Reply:
On https://www.postgresql.org/docs/current/sasl-authentication.html#SASL-SCRAM-SHA-256 it says that scram requires utf8 and postgres doesn't, which is the reason to do this saslprep. However, all `String`s in Crystal are UTF8, and there is no way to use a not-`String` to crystal-pg as it is. So maybe there is no reason to do this until (and if) non-string passwords are allowed?

Reply:
@will So what PostgreSQL does is that any UTF8 string goes through SASLprep. If it's not a UTF-8 string, or if it fails at certain parts of the SASLprep, then it just passes the string through. The server-side implementation can be found here:

https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/common/saslprep.c

Certain password will fail without following it. Here's an example of some of the test cases to try:

https://github.com/MagicStack/asyncpg/blob/master/tests/test_connect.py#L238

Reply:
moved this to https://github.com/will/crystal-pg/issues/181